### DISCUSSION TITLE
OpenSSL@3: Unable to use PKCS11 to sign a cert using a smartcard

### DISCUSSION THREAD

**Discussion by trelane** [Homebrew/discussions#5058] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1c612baa096c69f2fcb221c74e6f5b9979efdcee
Last commit: 18 hours ago
Core tap JSON: 16 Jan 07:51 UTC
Core cask tap JSON: 16 Jan 07:51 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.KtguBRI58t/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 16-core 64-bit arm_palma
Clang: 15.0.0 build 1500
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 12.5.0.22.9
Xcode: 15.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Command Line Tools are too outdated.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.


Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.


Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
```

### Description of issue

Getting a PCKS11 traceback when attempting to use a smart card.  I note that PKCS11 isn't listed as supported by OpenSSL.  I did install opensmartcard (opensc cask), pkcs11-tools, and pkcs11-helper but this doesn't help.  Redacted output is below.  OpenSSL@3 doesn't think it supports PKCS11.  Am I missing something?

$ openssl ca -config ./openssl_intermediate.cnf -engine pkcs11 -keyform engine -extensions server_cert -extensions req_ext -extfile REDACTED -days 1000 -notext -md sha256 -in REDACTED -out REDACTED
Invalid engine "pkcs11"
009043E401000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib): dlopen(/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib, 0x0002): tried: '/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file), '/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file)
009043E401000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:147:
009043E401000000:error:13000084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:438:
009043E401000000:error:13000074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:471:id=pkcs11
009043E401000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(libpkcs11.dylib): dlopen(libpkcs11.dylib, 0x0002): tried: 'libpkcs11.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibpkcs11.dylib' (no such file), '/usr/lib/libpkcs11.dylib' (no such file, not in dyld cache), 'libpkcs11.dylib' (no such file)
009043E401000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:147:
009043E401000000:error:13000084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:438:
Using configuration from ./openssl_intermediate.cnf
Error configuring OpenSSL modules
009043E401000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib): dlopen(/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib, 0x0002): tried: '/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file), '/opt/homebrew/Cellar/openssl@3/3.2.0_1/lib/engines-3/pkcs11.dylib' (no such file)
009043E401000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:147:
009043E401000000:error:13000084:engine routines:dynamic_load:dso not found:crypto/engine/eng_dyn.c:438:
009043E401000000:error:13000074:engine routines:ENGINE_by_id:no such engine:crypto/engine/eng_list.c:471:id=pkcs11
009043E401000000:error:13000066:engine routines:int_engine_configure:engine configuration error:crypto/engine/eng_cnf.c:139:section=pkcs11_section, name=init, value=0
009043E401000000:error:0700006D:configuration file routines:module_run:module initialization error:crypto/conf/conf_mod.c:276:module=engines, value=engine_section retcode=-1
trelane@MacBook-Pro Secure_Key % openssl
help:

Standard commands
asn1parse         ca                ciphers           cmp
cms               crl               crl2pkcs7         dgst
dhparam           dsa               dsaparam          ec
ecparam           enc               engine            errstr
fipsinstall       gendsa            genpkey           genrsa
help              info              kdf               list
mac               nseq              ocsp              passwd
pkcs12            pkcs7             pkcs8             pkey
pkeyparam         pkeyutl           prime             rand
rehash            req               rsa               rsautl
s_client          s_server          s_time            sess_id
smime             speed             spkac             srp
storeutl          ts                verify            version
x509

Message Digest commands (see the `dgst' command for more details)
blake2b512        blake2s256        md4               md5
mdc2              rmd160            sha1              sha224
sha256            sha3-224          sha3-256          sha3-384
sha3-512          sha384            sha512            sha512-224
sha512-256        shake128          shake256          sm3

Cipher commands (see the `enc' command for more details)
aes-128-cbc       aes-128-ecb       aes-192-cbc       aes-192-ecb
aes-256-cbc       aes-256-ecb       aria-128-cbc      aria-128-cfb
aria-128-cfb1     aria-128-cfb8     aria-128-ctr      aria-128-ecb
aria-128-ofb      aria-192-cbc      aria-192-cfb      aria-192-cfb1
aria-192-cfb8     aria-192-ctr      aria-192-ecb      aria-192-ofb
aria-256-cbc      aria-256-cfb      aria-256-cfb1     aria-256-cfb8
aria-256-ctr      aria-256-ecb      aria-256-ofb      base64
bf                bf-cbc            bf-cfb            bf-ecb
bf-ofb            camellia-128-cbc  camellia-128-ecb  camellia-192-cbc
camellia-192-ecb  camellia-256-cbc  camellia-256-ecb  cast
cast-cbc          cast5-cbc         cast5-cfb         cast5-ecb
cast5-ofb         des               des-cbc           des-cfb
des-ecb           des-ede           des-ede-cbc       des-ede-cfb
des-ede-ofb       des-ede3          des-ede3-cbc      des-ede3-cfb
des-ede3-ofb      des-ofb           des3              desx
idea              idea-cbc          idea-cfb          idea-ecb
idea-ofb          rc2               rc2-40-cbc        rc2-64-cbc
rc2-cbc           rc2-cfb           rc2-ecb           rc2-ofb
rc4               rc4-40            seed              seed-cbc
seed-cfb          seed-ecb          seed-ofb          sm4-cbc
sm4-cfb           sm4-ctr           sm4-ecb           sm4-ofb

trelane@MacBook-Pro Secure_Key % openssl --version
OpenSSL 3.2.0 23 Nov 2023 (Library: OpenSSL 3.2.0 23 Nov 2023)

### Relevant casks

openssl@3 opensc pcks11-helper pkcs11-tools

**Comment by Dreamacro** [comment]
Temporary workaround

* install `libp11` `brew install libp11`
* ln `pkcs11.dylib` manually
    *  `cd /opt/homebrew/Cellar/openssl@3/3.3.0/lib/engines-3`
    * `ln -s /opt/homebrew/Cellar/libp11/0.4.12/lib/engines-3/pkcs11.dylib .`
* verify `openssl engine pkcs11 -t`
